# Machine Learning Challenge - Round -1

# Imports

In [53]:
%load_ext autoreload
%autoreload 2

%matplotlib inline

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [54]:
from fastai.imports import *
from fastai.structured import *

from pandas_summary import DataFrameSummary
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from IPython.display import display

from sklearn import metrics

Path for the data

In [55]:
PATH = "data/yesbank/"

In [56]:
df_raw = pd.read_csv(f'{PATH}ml2_train.csv', low_memory=False, 
                     parse_dates=["Established_Date"],dayfirst = True)

df_test = pd.read_csv(f'{PATH}ml2_test.csv', low_memory=False,
                    parse_dates = ["Established_Date"],dayfirst = True)

it's important to look at your data, to make sure you understand the format, how it's stored, what type of values it holds, etc.

In [57]:
def display_all(df):
    with pd.option_context("display.max_rows", 1000, "display.max_columns", 1000):
        display(df)

In [58]:
display_all(df_raw.tail().T)

3785                 3786  \
Serial_Number                    3786                 3787   
Main_Office                         0                    0   
Branch_Number                    7981                 7982   
Established_Date  2015-03-11 00:00:00  2016-02-02 00:00:00   
Acquired_Date                     NaN                  NaN   
City                          Compton            Las Vegas   
County                    Los Angeles                Clark   
State                              CA                   NV   
Deposits_2010                     NaN                  NaN   
Deposits_2011                     NaN                  NaN   
Deposits_2012                     NaN                  NaN   
Deposits_2013                     NaN                  NaN   
Deposits_2014                     NaN                  NaN   
Deposits_2015                     NaN                  NaN   
Deposits_2016                    3446                 2666   

                                 3787                 3788  \
Serial_Number                    3788                 3789   
Main_Office                         0                    0   
Branch_Number                    7984                 7988   
Established_Date  2016-03-15 00:00:00  2016-01-01 00:00:00   
Acquired_Date                     NaN                  NaN   
City                           Irvine          New Orleans   
County                         Orange              Orleans   
State                              CA                   LA   
Deposits_2010                     NaN                  NaN   
Deposits_2011                     NaN                  NaN   
Deposits_2012                     NaN                  NaN   
Deposits_2013                     NaN                  NaN   
Deposits_2014                     NaN                  NaN   
Deposits_2015                     NaN                  NaN   
Deposits_2016                    6689                94133   

                                 3789  
Serial_Number                    3790  
Main_Office                         0  
Branch_Number                    7989  
Established_Date  2016-01-01 00:00:00  
Acquired_Date                     NaN  
City                          Buffalo  
County                           Erie  
State                              NY  
Deposits_2010                     NaN  
Deposits_2011                     NaN  
Deposits_2012                     NaN  
Deposits_2013                     NaN  
Deposits_2014                     NaN  
Deposits_2015                     NaN  
Deposits_2016                   45596

In [59]:
display_all(df_raw.describe(include='all').T)

count unique                  top  freq                first  \
Serial_Number     3790    NaN                  NaN   NaN                  NaN   
Main_Office       3790    NaN                  NaN   NaN                  NaN   
Branch_Number     3790    NaN                  NaN   NaN                  NaN   
Established_Date  3790   2027  1919-01-01 00:00:00   129  1825-01-01 00:00:00   
Acquired_Date     2297      5           09/25/2008  1528                  NaN   
City              3790   1489              Chicago    99                  NaN   
County            3790    299          Los Angeles   298                  NaN   
State             3790     25                   CA  1003                  NaN   
Deposits_2010     3050    NaN                  NaN   NaN                  NaN   
Deposits_2011     3212    NaN                  NaN   NaN                  NaN   
Deposits_2012     3461    NaN                  NaN   NaN                  NaN   
Deposits_2013     3615    NaN                  NaN   NaN                  NaN   
Deposits_2014     3734    NaN                  NaN   NaN                  NaN   
Deposits_2015     3771    NaN                  NaN   NaN                  NaN   
Deposits_2016     3790    NaN                  NaN   NaN                  NaN   

                                 last     mean          std   min      25%  \
Serial_Number                     NaN   1895.5      1094.22     1   948.25   
Main_Office                       NaN        0            0     0        0   
Branch_Number                     NaN  5390.81      1549.11  2866  4062.25   
Established_Date  2016-10-05 00:00:00      NaN          NaN   NaN      NaN   
Acquired_Date                     NaN      NaN          NaN   NaN      NaN   
City                              NaN      NaN          NaN   NaN      NaN   
County                            NaN      NaN          NaN   NaN      NaN   
State                             NaN      NaN          NaN   NaN      NaN   
Deposits_2010                     NaN   112213  2.79759e+06     0    18932   
Deposits_2011                     NaN   125514  3.44738e+06     0  20132.8   
Deposits_2012                     NaN   128920  3.54715e+06     0    21065   
Deposits_2013                     NaN   136383  3.74036e+06     0  23314.5   
Deposits_2014                     NaN   145592  3.98028e+06     0    26239   
Deposits_2015                     NaN   157628  4.28141e+06     0    30881   
Deposits_2016                     NaN   183256  5.11612e+06     0  37091.5   

                     50%      75%          max  
Serial_Number     1895.5  2842.75         3790  
Main_Office            0        0            0  
Branch_Number     5256.5  6858.25         7989  
Established_Date     NaN      NaN          NaN  
Acquired_Date        NaN      NaN          NaN  
City                 NaN      NaN          NaN  
County               NaN      NaN          NaN  
State                NaN      NaN          NaN  
Deposits_2010      35628  66072.8  1.53577e+08  
Deposits_2011      37183    66947  1.94388e+08  
Deposits_2012      39744    71496  2.07368e+08  
Deposits_2013      42358  76352.5  2.23395e+08  
Deposits_2014      46772  83296.5  2.41541e+08  
Deposits_2015      52516    91566  2.61293e+08  
Deposits_2016      60753   104378  3.13687e+08

# Initial Processing

This dataset contains a mix of continuous and categorical variables.
The below method extracts particular date fields from a complete datetime for the purpose of constructing categoricals.

In [60]:
add_datepart(df_raw,'Established_Date')
df_raw.Established_Year.head()
add_datepart(df_test,'Established_Date')

we call the function train_cats to convert strings to pandas categories.

In [61]:
train_cats(df_raw)
train_cats(df_test)

The below step is to replace the text to numericals

In [62]:
df_raw.City.cat.categories
df_raw.County.cat.categories
df_raw.State.cat.categories
df_test.City.cat.categories
df_test.County.cat.categories
df_test.State.cat.categories

Index(['AZ', 'CT', 'FL', 'IL', 'KY', 'LA', 'MI', 'NJ', 'NY', 'OH', 'TX', 'UT',
       'WI', 'WV'],
      dtype='object')

output : Index(['AZ', 'CT', 'FL', 'IL', 'KY', 'LA', 'MI', 'NJ', 'NY', 'OH', 'TX', 'UT',
       'WI', 'WV'],
      dtype='object')

In [63]:
df_raw.City = df_raw.City.cat.codes
df_raw.County = df_raw.County.cat.codes
df_raw.State = df_raw.State.cat.codes
df_test.City = df_test.City.cat.codes
df_test.County = df_test.County.cat.codes
df_test.State = df_test.State.cat.codes

Since we have lots of missing values, which we can't pass directly to a random forest.

In [64]:
display_all(df_raw.isnull().sum().sort_index()/len(df_raw))
display_all(df_test.isnull().sum().sort_index()/len(df_test))

Acquired_Date                   0.393931
Branch_Number                   0.000000
City                            0.000000
County                          0.000000
Deposits_2010                   0.195251
Deposits_2011                   0.152507
Deposits_2012                   0.086807
Deposits_2013                   0.046174
Deposits_2014                   0.014776
Deposits_2015                   0.005013
Deposits_2016                   0.000000
Established_Day                 0.000000
Established_Dayofweek           0.000000
Established_Dayofyear           0.000000
Established_Elapsed             0.000000
Established_Is_month_end        0.000000
Established_Is_month_start      0.000000
Established_Is_quarter_end      0.000000
Established_Is_quarter_start    0.000000
Established_Is_year_end         0.000000
Established_Is_year_start       0.000000
Established_Month               0.000000
Established_Week                0.000000
Established_Year                0.000000
Main_Office     

Acquired_Date                   0.075169
Branch_Number                   0.000000
City                            0.000000
County                          0.000000
Deposits_2010                   0.000000
Deposits_2011                   0.000000
Deposits_2012                   0.000000
Deposits_2013                   0.000000
Deposits_2014                   0.000000
Deposits_2015                   0.000000
Established_Day                 0.000000
Established_Dayofweek           0.000000
Established_Dayofyear           0.000000
Established_Elapsed             0.000000
Established_Is_month_end        0.000000
Established_Is_month_start      0.000000
Established_Is_quarter_end      0.000000
Established_Is_quarter_start    0.000000
Established_Is_year_end         0.000000
Established_Is_year_start       0.000000
Established_Month               0.000000
Established_Week                0.000000
Established_Year                0.000000
Main_Office                     0.000000
Serial_Number   

# Preprocessing

Since we have lots of training data compared to test data available, the training data has been splitted to required number in order to overcome overfitting.

In [39]:
def split_vals(a,n): return a[:n].copy(), a[n:].copy()

n_valid = 2167  # same as Kaggle's test set size
n_trn = len(df_raw)-n_valid
raw_train, raw_valid = split_vals(df_raw, n_trn)
raw_train.shape
df_raw = raw_train

We'll replace categories with their numeric codes, handle missing continuous values, and split the dependent variable into a separate variable.

In [66]:
df, y, nas = proc_df(df_raw, 'Deposits_2016')
df1, y1, nas1 = proc_df(df_test)
df

Serial_Number  Main_Office  Branch_Number  Acquired_Date  City  County  \
0                 1            0           2866              5  1383     279   
1                 2            0           2867              5   451     277   
2                 3            0           2868              5   137     279   
3                 4            0           2869              5  1005     279   
4                 5            0           2870              5  1401     279   
5                 6            0           2871              5  1401     279   
6                 7            0           2872              5   137     279   
7                 8            0           2873              5   859     279   
8                 9            0           2874              5   936     279   
9                10            0           2875              5   159     279   
10               11            0           2876              5   845     279   
11               12            0           2877              5  1401     279   
12               13            0           2878              5   428     169   
13               14            0           2879              5   795     169   
14               15            0           2880              5   795     169   
15               16            0           2881              5   408      83   
16               17            0           2883              5   959     292   
17               18            0           2884              5   959     292   
18               19            0           2885              5  1065     211   
19               20            0           2886              5  1065     211   
20               21            0           2887              5  1065     211   
21               22            0           2888              5   603     126   
22               23            0           2889              5  1025     126   
23               24            0           2890              5   132     126   
24               25            0           2891              5   603     126   
25               26            0           2892              5   603     126   
26               27            0           2893              5  1270     207   
27               28            0           2894              5    24     139   
28               29            0           2895              5    28     193   
29               30            0           2896              5    28     292   
...             ...          ...            ...            ...   ...     ...   
3760           3761            0           7952              0   105     194   
3761           3762            0           7953              0  1432     190   
3762           3763            0           7955              0   972     194   
3763           3764            0           7956              0   579      70   
3764           3765            0           7957              0   458      49   
3765           3766            0           7958              0    74     285   
3766           3767            0           7959              0  1185     232   
3767           3768            0           7960              0  1249     227   
3768           3769            0           7961              0   105     194   
3769           3770            0           7962              0   917      80   
3770           3771            0           7963              0   398      39   
3771           3772            0           7964              0   118     154   
3772           3773            0           7965              0   790     156   
3773           3774            0           7967              0   399     112   
3774           3775            0           7969              0   253     166   
3775           3776            0           7970              0  1406     194   
3776           3777            0           7971              0   195       4   
3777           3778            0           7972              0  1002     199   
3778  

Now, the data is ready to fit the model using random forest regressor

In [67]:
m = RandomForestRegressor(n_jobs=-1)
m.fit(df, y)
m.score(df,y)

/home/paperspace/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/ensemble/forest.py:248: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9206251464882599

Output: 0.9206251464882599

The output is the coefficient of R^2

This shows that the train data given is efficient for the prediction process of the test data.

## Prediction of Test data 

In [17]:
deposit = m.predict(df1)
deposit

array([1.65417e+08, 3.58472e+05, 2.86499e+05, ..., 1.06214e+05, 8.45572e+04, 2.75120e+04])

## Clustering

In [18]:
A=[];
B=[];
C=[];
S1=[];
S2=[];
S3=[];
clu_a =[];
for i in range(len(deposit)):
    if (deposit[i]>=20000 and deposit[i]<=90000):
        A.append(deposit[i])
        S1.append(i+1)
    if (deposit[i] < 20000):
        B.append(deposit[i])
        S2.append(i+1)
    if (deposit[i] >90000):
        C.append(deposit[i])
        S3.append(i+1)    
deposit = A + B + C
Serialnum = S1 + S2 + S3



## Creation of csv file

In [20]:
import csv

with open('submissionlast.csv','w',newline = '') as f:
    thewriter = csv.writer(f)
    thewriter.writerow(['Serial Number','2016 Deposits'])
    for i in range(len(deposit)):
        thewriter.writerow([Serialnum[i],deposit[i]])
